### PR TITLE
Miner rate limit

### PR DESCRIPTION
**Pull request by Vovchyk**
## Description
This change introduces rate limit for miner submissions. By default, it's only enabled for `testnet`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

* **Other information**:

### CODE REVIEW

File: rskj-core/src/main/resources/config/testnet.conf [hunk 1]
```diff
@@ -46,6 +46,10 @@ peer {
     networkId = 8100
 }
 
+# the number of milliseconds that should pass from a previous submission of a mining solution before next one is allowed
+# there's no limit, if the number is zero or negative
```
Comment (nagarev) [inline rskj-core/src/main/resources/config/testnet.conf:50]: I would change it to: "Set this value to zero or negative number to disable this limit"

File: rskj-core/src/main/resources/reference.conf [hunk 2]
```diff
@@ -130,6 +130,10 @@ miner {
     server {
         enabled = false
         isFixedClock = false
+
+        # the number of milliseconds that should pass from a previous submission of a mining solution before next one is allowed
+        # there's no limit, if the number is zero or negative
```
Comment (nagarev) [inline rskj-core/src/main/resources/reference.conf:135]: Same comment as in `testnet.conf`

File: rskj-core/src/main/resources/config/testnet.conf (hunk 1 above)
Comment (Vovchyk) [reply rskj-core/src/main/resources/config/testnet.conf]: done

File: rskj-core/src/main/resources/reference.conf (hunk 2 above)
Comment (Vovchyk) [reply rskj-core/src/main/resources/reference.conf]: done

File: rskj-core/src/test/java/co/rsk/mine/MinerServerTest.java [hunk 3]
```diff
@@ -923,4 +692,73 @@ private void findNonce(MinerWork work, BtcBlock bitcoinMergedMiningBlock) {
             }
         }
     }
+
+    private MinerServerImpl makeMinerServer(EthereumImpl ethereumImpl, BlockUnclesValidationRule unclesValidationRule,
```
Comment (nagarev) [inline rskj-core/src/test/java/co/rsk/mine/MinerServerTest.java:696]: Beatiful refactorings in the whole file! Only a suggestion: `makeMinerServer` can be called `makeMinerServerMock`, I think it easier to understand what's happening while reading the method invocation. As the other suggestions, is only an opinion, feel free to dismiss it if you don't like it.

File: rskj-core/src/main/java/co/rsk/mine/MinerServerImpl.java [hunk 4]
```diff
@@ -260,7 +278,18 @@ private SubmitBlockResult processSolution(
...
+            if (submissionRateLimitHandler.isEnabled()) {
+                if (!submissionRateLimitHandler.isSubmissionAllowed()) {
+                    String message = "Cannot publish block, block submission rate limit exceeded";
+                    logger.warn(message);
+
+                    return new SubmitBlockResult("ERROR", message);
+                }
+
+                submissionRateLimitHandler.onSubmit();
+            }
```
[the remark this answers is not in the document]
Comment (Vovchyk) [reply rskj-core/src/main/java/co/rsk/mine/MinerServerImpl.java:290]: thanks for the suggestion. But these are just a few lines of code, and the logging part is tied to the `MinerServerImpl` class. Personally I think that this code snippet is more related to the `MinerServerImpl` class

File: rskj-core/src/test/java/co/rsk/mine/MinerServerTest.java (hunk 3 above)
Comment (nagarev) [reply rskj-core/src/test/java/co/rsk/mine/MinerServerTest.java:696]: Excellent Idea! we can add it to the readme, and establish some procedure like "When doing or modifying some test on an existent class, we should try to update as many names as possible to comply with the naming convention" or something like that.

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Comment by Vovchyk** [comment]
pipeline:run

**Review by heitorfm — APPROVED** [review]

**Pull request merged**